lib/semantics/check-select-stmt.cpp

```diff
@@ -0,0 +1,316 @@
+//===-- lib/semantics/check-select-stmt.cc --------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "check-select-stmt.h"
+#include "flang/parser/message.h"
+#include "flang/semantics/tools.h"
+
+namespace Fortran::semantics {
+
+class SelectCaseHelper {
+
+public:
+  SelectCaseHelper(parser::Messages &messages) : messages_{messages} {}
+
+  void ReportOverlapCaseValError(
+      parser::CharBlock &source1, parser::CharBlock &source2) {
+    // C1149
+    messages_
+        .Say(source2,
+            "SELECT CASE statement value at this location overlaps with the below location"_err_en_US)
+        .Attach(source1,
+            "SELECT CASE statement value overlaps with the above location"_err_en_US);
+  }
+
+  void Insert(std::int64_t intVal, parser::CharBlock source) {
+    auto selectCaseIntRangeValsIter =
+        std::find_if(selectCaseIntRangeVals_.begin(),
+            selectCaseIntRangeVals_.end(), [&intVal](const auto &mem) {
+              return ((!mem.fromVal_ || (intVal >= mem.fromVal_.value())) &&
+                  (!mem.toVal_ || (intVal <= mem.toVal_.value())));
+            });
+    if (selectCaseIntRangeValsIter != selectCaseIntRangeVals_.end()) {
+      ReportOverlapCaseValError(selectCaseIntRangeValsIter->source_, source);
+    }
+    auto selectCaseIntValsInsertStatus =
+        selectCaseIntVals_.insert(std::make_pair(intVal, source));
+    if (selectCaseIntValsInsertStatus.second == false) {
+      ReportOverlapCaseValError(
+          selectCaseIntValsInsertStatus.first->second, source);
+    }
+  }
+  void Insert(std::string strVal, parser::CharBlock source) {
```

Comment:
All of my comments from the previous function apply to this one, and probably the one after.  Since they look like they're nearly identical, I think that you could abstract them into a template and then you'd have to fix the problems only once.

Reply:
As suggested by you and Tim, I will abstract out the common code and keep it separate as a function.